### ISSUE TITLE
TFLite slower with NNAPI on Snapdragon 660

### ISSUE THREAD

**Issue by alexcohn** [tensorflow/tensorflow#28283]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **No**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **n/a**
- Mobile device: **Xiaomi Mi A2**, Android 9
- TensorFlow installed from: **binary** 
- TensorFlow version: **1.13.1** 
- GPU model and memory: **Adreno 512**

**Describe the current behavior**

Running the [TFLite demo](https://github.com/tensorflow/tensorflow/tree/6d28416/tensorflow/lite/java/demo) (the link is the HEAD of master branch as of now) on , average inference time for 1 thread, _mobilenet v1 quant_, on CPU is **70 ms** and with NNAPI degrades to **170 ms**. For _mobilenet v1 float_, CPU is around **150 ms**, and NNAPI degrades to **800 ms**. Choosing GPU (only possible for float model) delivers same **70 ms** as CPU/quant.

On the other hand, a weaker **Xiaomi Mi A1** device (which got official upgrade to Android 9), NNAPI does provide x2 acceleration (from around **140 ms** for quantized net to **70 ms**).

**Describe the expected behavior**

I would expect NNAPI at least not to cause performance degradation. If I understand correctly, the library is smart enough to choose between GPU and CPU.

**Other info / logs**

A1: 

- Qualcomm MSM8953 Snapdragon 625 
- Octa-core 2.0 GHz Cortex-A53 
- Adreno 506

A2: 

- Qualcomm SDM660 Snapdragon 660 
- Octa-core (4x2.2 GHz Kryo 260 & 4x1.8 GHz Kryo 260) 
- Adreno 512

*This is similar to other devices, e.g. https://github.com/tensorflow/tensorflow/issues/15554*

**Comment by silfverstrom** [comment]
I'm experiencing the same problem using an Xiaomi Mi 9(Snapdragon 855). 

I'm experiencing significant inference slowdown for some models when using the NNApi. 
I've benchmarked using TFLite Model Benchmark, and tried a couple of models from https://www.tensorflow.org/lite/guide/hosted_models

Densenet, Resnet and NASNet are much slower using the NNAPI, while MobileNet and Inception(V3, V4) are much faster.

**Comment by freedomtan** [comment]
@alexcohn as you can see in the name, NNAPI is an API. What you get when you call NNAPI depends on underlying implementation(s). If the device you use doesn't have NNAPI driver(s), NNAPI's cpu fallback will be used, see [NNAPI doc](https://developer.android.com/ndk/guides/neuralnetworks). NNAPI cpu fallback is not a part of TensorFlow Lite. It's a part of AOSP. That is, this is unlikely a TensorFlow issue. It seems you ran into either NNAPI fallback or bad drivers. If you take a look at TFLite's and NNAPI's source code, you can see that they actually share the same code base for kernels. But they not always synced because NNAPI release time has to be aligned with Android's release time. And there are no way to set something like the number of threads for CPU fallback in NNAPI. Thus, you see performance mismatch.

**Comment by alexcohn** [comment]
@freedomtan Thanks for your answer. 

So, the conclusion is that my app should include the 'discovery logic'. That is, after install, it must benchmark my specific model for both NNAPI and CPU (and also, the GPU delegate), and decide which delivers best performance on the particular device.

I also came across the [Neural Network Transpiler](https://github.com/alexst07/neural-network-transpiler) project that purports to *Convert a model from tflite to a C++ source code using Android Neural Network API*. I am kind of sceptical now after reading your answer. Is there any chance that such static conversion will provide reliably better performance, compared to TFLite CPU or NNAPI option?

**Comment by freedomtan** [comment]
@alexcohn Ideally, NNAPI should do it for you. However, as far as I can tell, it's not there yet. 

Regarding the Neural Network Transpiler, if it is done right, it should be able to reduce some memory management overhead.

**Comment by alanchiao** [comment]
Just coming here to confirm "freedom"'s answer. 

NNAPI updates as a part of the yearly Android release cycle, including any CPU optimizations. They share the same performance-impacting CPU code.  

So when comparing NNAPI versus TFLite, there are a few factors:
- NNAPI is faster due to hardware acceleration outside the CPU
- NNAPI is slower on the CPU since it lags TFLite's optimizations by up to a year. 
     - This also includes the fact that while NNAPI provides speedups for the hosted models page ("fully quantized models"), it doesn't for models generated by this [tooling]
        (https://www.tensorflow.org/lite/performance/post_training_quantization)("hybrid or weight-only quantized models").

It's good that you saw the speedups on the Xiaomi. For the other device, (I guess the A1?), it sounds like it could have an even older version of Android / NNAPI (Android 8?).

**Comment by alexcohn** [comment]
No, I did not see speedup on the newer Xiaomi A2, running with official Android 9. The x2 speedup with NNAPI I saw was on the older Xiaomi A1 device, which recently received an official OTA upgrade to Android 9.

I totally understand that release cycle of NNAPI cannot keep at pace with FTLite. My major disappointment was that it does not seem possible to predict whether using NNAPI with either `setUseNNAPI(true)` or with `addDelegate(NnApiDelegate())` will deliver a substantial improvement or an even more substantial slowdown.

**Comment by jdduke** [comment]
Hey @alexcohn, we're working on some tooling to make it easier to predict/understand NNAPI performance for a given device+model combination. Historically we've relied on vendors to vet the performance of their driver, but there can be surprising performance cliffs in certain cases. Thanks for the feedback; it's definitely an issue that's on our radar and something we want to solve generally to make it easy to hit the fast paths.

**Comment by Noltibus** [comment]
I faced the same difficulties having slower inference time when activating the NNAPI on different devices (Google Pixel XL, Google Pixel 3XL, Samsung Galaxy S8, Motorola Moto G5S Plus). Also, I tried to use [this](http://ai-benchmark.com/ranking_processors.html) to check if there might be a connection between having an AI Accelerator as it is called in the ranking and profiting from turning NNAPI on. But according to the ranking, the Pixel 3 XL should profit from the NNAPI but my tensorflow lite model runs in 48 ms compared to 31 ms without NNAPI.

**Comment by silfverstrom** [comment]
@Noltibus It must likely is a problem with your model. Have you checked if your ops in your model are compatible with the NNAPI? https://developer.android.com/ndk/guides/neuralnetworks#operations

When you have ops in your graph that are not compatible with the NNAPI, the CPU-fallback is used which will cause multiple graphs to be executed, which in turns causes some major slowdowns.

**Comment by Noltibus** [comment]
This is my model shown in Netron:
![smaller_alex_lite](https://user-images.githubusercontent.com/15614838/59684024-7ba74000-91d9-11e9-8496-b09949670d23.png)
The only operation I cannot find in the reference is the SPLIT operation, but is that really the bottleneck?

**Comment by silfverstrom** [comment]
@Noltibus Try removing the split and see if the inference-time using the NNAPI improves relative to CPU. I would guess that the split causes your graph to be divided into two subgraphs, which would slow down inference quite a bit. 

Another thing to take into consideration is that hexagons DSP only works in [8bit.](https://developer.qualcomm.com/docs/snpe/quantized_models.html) If you use a floating point model the network initialization time will increase as the DSP automatically quantize the network parameters. In my experience, DSP-inference on smaller models(smaller inputs, fewer parameters) won't outperform a CPU counterpart.

**Comment by Noltibus** [comment]
I will try and report back! Thank you for the input!

**Comment by alexcohn** [comment]
I wonder whether there are mainstream devices where NNAPI is faster than GPU delegate. The catch is that the list of ops implemented in GPU is limited. But AFAIK, the list of ops implemented in NNAPI is even shorter.

**Comment by freedomtan** [comment]
@alexcohn Definitely there are devices with good NN accelerators and corresponding NN HAL drivers. You may want to check the list @Noltibus mentioned. On my Pixel 3a, running Android Q Beta 4, MobileNet V1 takes around 55 ms (fp16) with GPU delegate and less than 8 ms (quantized int8) with NNAPI delegate to DSP.

**Comment by aiff22** [comment]
Hi @alexcohn,

You can take a look at http://ai-benchmark.com/ranking_detailed to see the speed of different NN models running with NNAPI on various Android devices. For float and quantized models, you should see a huge speed-up on all phones with SDM855, Kirin 980/810 and Helio P90.

**Comment by alexcohn** [comment]
@aiff22, I have studied this table a lot, and I don't doubt that some devices have nice acceleration via NNAPI. The issue is that for some devices, same NNAPI makes the inference *slower*, so we cannot use it *everywhere*. Another question, is NNAPI always better than GPU delegate on devices with hardware acceleration?

**Comment by aiff22** [comment]
@alexcohn, yes, right now NNAPI is dramatically better than GPU delegate on devices with hardware acceleration. Besides that, the current version of GPU delegate contains a huge number of critical bugs, thus I would recommend using it **really carefully** and always check the actual outputs that are produced by the model (e.g. NANs are returned for almost any model running on Mali GPUs).

Regarding NNAPI - it won't give you better results if vendor's NNAPI drivers are not present in Android image. It makes sense always use NNAPI only for the following chipsets:

- Snapdragon 855 / 730
- Helio P90
- Kirin 980 / 970 / 810
- Exynos 9820

If you have a quantized model, you can enable NNAPI for the next SoCs as well (even if vendor's NNAPI drivers are not present, the results won't be much worse):

- Snapdragon 845 / 712 / 710 / 675 / 670 / 665
- Helio P70 / P60

In all other cases just use TFLite with CPU and set the number of threads to min(4, number_of_cpu_cores). This will guarantee you the fastest execution speed and the absence of accuracy problems.

**Comment by alexcohn** [comment]
Thanks, Andrey. I believe that your answer can close this issue.

**Comment by tgpsantos** [comment]
I have an Exynos 9810 with Mali G72 and I've never got a NaN (not that I had seen at least). But tbh I never saw an increase in performance when switching the interpreter from cpu to gpu

**Comment by tgpsantos** [comment]
> 
> 
> @alexcohn, yes, right now NNAPI is dramatically better than GPU delegate on devices with hardware acceleration. Besides that, the current version of GPU delegate contains a huge number of critical bugs, thus I would recommend using it **really carefully** and always check the actual outputs that are produced by the model (e.g. NANs are returned for almost any model running on Mali GPUs).
> 
> Regarding NNAPI - it won't give you better results if vendor's NNAPI drivers are not present in Android image. It makes sense always use NNAPI only for the following chipsets:
> 
>     * Snapdragon 855 / 730
> 
>     * Helio P90
> 
>     * Kirin 980 / 970 / 810
> 
>     * Exynos 9820
> 
> 
> If you have a quantized model, you can enable NNAPI for the next SoCs as well (even if vendor's NNAPI drivers are not present, the results won't be much worse):
> 
>     * Snapdragon 845 / 712 / 710 / 675 / 670 / 665
> 
>     * Helio P70 / P60
> 
> 
> In all other cases just use TFLite with CPU and set the number of threads to min(4, number_of_cpu_cores). This will guarantee you the fastest execution speed and the absence of accuracy problems.

I have an Exynos 9810 with Mali G72 and I never got any NaN (not that I had noticed at least). But tbh I never saw an increase in performance when switching the interpreter from cpu to gpu so I even doubt if that's actually running on gpu(?)

**Comment by aiff22** [comment]
@tgpsantos, if runtime is the same - this usually means that the model contains some unsupported ops, and thus CPU backend is used instead of GPU.

**Comment by allenling** [comment]
@aiff22 how could i check there is any unsupported op in my tflite model?

**Comment by allenling** [comment]
@aiff22 also, when i run tflite object detection [example](https://github.com/tensorflow/examples/tree/master/lite/examples/object_detection/android) in Huawei P20, which equipped with an Kirin 970 CPU, enable nnapi will make inference slower(from 40ms to 200ms)

**Comment by silfverstrom** [comment]
@allenling I would recommend that you try the benchmark tool found at,
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/benchmark

You can find a list of supported ops at https://developer.android.com/ndk/guides/neuralnetworks#operands

Note that for quantized models, as per the example, there are a couple of limitations in NNAPI <= 1.1. For instance, [hybrid operators](https://www.tensorflow.org/lite/performance/post_training_quantization) are not delegated to the NNAPI, bilinear interpolation only works when width == height etc. So there are a couple of limitations that could make the execution graph split up into multiple sub-graphs, some executed on an accelerator and some on CPU. 

Best way to check this would be to enable --enable_op_profiling=true when running the benchmark util mentioned above.

**Comment by Noltibus** [comment]
Coming back to this issue about 6 months later: 

> The only operation I cannot find in the reference is the SPLIT operation, but is that really the bottleneck? 

Now, according to the website, the SPLIT operation is supported, but I still have a worse performance with my Pixel 3 XL running Android 10 and using NNApi compared to not using it. On a Pixel XL running Android 10, the peformance with and without the NNApi is the same. This still sounds like there is an unsupported operation, right? But is the [SPLIT](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0afb787314a8631fe847f1fd93cfd576a7) operation mentioned in the docs the right one?

**Comment by Noltibus** [comment]
> @alexcohn, yes, right now NNAPI is dramatically better than GPU delegate on devices with hardware acceleration. Besides that, the current version of GPU delegate contains a huge number of critical bugs, thus I would recommend using it **really carefully** and always check the actual outputs that are produced by the model (e.g. NANs are returned for almost any model running on Mali GPUs).
> 
> Regarding NNAPI - it won't give you better results if vendor's NNAPI drivers are not present in Android image. It makes sense always use NNAPI only for the following chipsets:
> 
> * Snapdragon 855 / 730
> * Helio P90
> * Kirin 980 / 970 / 810
> * Exynos 9820
> 
> If you have a quantized model, you can enable NNAPI for the next SoCs as well (even if vendor's NNAPI drivers are not present, the results won't be much worse):
> 
> * Snapdragon 845 / 712 / 710 / 675 / 670 / 665
> * Helio P70 / P60
> 
> In all other cases just use TFLite with CPU and set the number of threads to min(4, number_of_cpu_cores). This will guarantee you the fastest execution speed and the absence of accuracy problems.

There is no reliable way of finding out if the phone the app is running on has one of these processors, right?

**Comment by jdduke** [comment]
Keep in mind that even if NNAPI states that a given op is supported, hardware drivers for NNAPI may not support that op, or all types for that op (e.g., the Pixel 3 has a DSP driver for quantized models, but may not accelerate float models).

We're aware of the complexity in predicting how well NNAPI will support a model across the wide permutation of drivers and devices, and are working on some tweaks to the delegate to both avoid regressing performance and help whitelist the fast paths for model/device combinations. This is a major priority for the team over the next few months.

**Comment by alexcohn** [comment]
> There is no reliable way of finding out if the phone the app is running on has one of these processors, right?

Reliable, yes. Officially documented, probably no.
